File: simplified-app-shared/src/main/java/org/nypl/simplified/app/catalog/CatalogBookRevokeButton.java
package org.nypl.simplified.app.catalog;

import android.content.res.Resources;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.AppCompatActivity;

import com.io7m.jnull.NullCheck;

import org.nypl.simplified.app.R;
import org.nypl.simplified.books.accounts.AccountType;
import org.nypl.simplified.books.book_database.BookID;
import org.nypl.simplified.books.controller.BooksControllerType;

/**
 * A button for revoking loans or holds.
 */

public final class CatalogBookRevokeButton
  extends CatalogLeftPaddedButton implements CatalogBookButtonType {

  /**
   * Construct a button.
   */

  public CatalogBookRevokeButton(
    final AppCompatActivity in_activity,
    final BooksControllerType in_books,
    final AccountType in_account,
    final BookID in_book_id,
    final CatalogBookRevokeType in_revoke_type) {

    super(in_activity);

    NullCheck.notNull(in_book_id);
    NullCheck.notNull(in_revoke_type);

    final Resources resources = NullCheck.notNull(in_activity.getResources());

    switch (in_revoke_type) {
      case REVOKE_LOAN: {
        this.getTextView().setText(NullCheck.notNull(resources.getString(R.string.catalog_book_revoke_loan)));
        this.setContentDescription(NullCheck.notNull(resources.getString(R.string.catalog_accessibility_book_revoke_loan)));
        break;
      }
      case REVOKE_HOLD: {
        this.getTextView().setText(NullCheck.notNull(resources.getString(R.string.catalog_book_revoke_hold)));
        this.setContentDescription(NullCheck.notNull(resources.getString(R.string.catalog_accessibility_book_revoke_hold)));
        break;
      }
    }

    this.setOnClickListener(view -> {
      final CatalogBookRevokeDialog d =
        CatalogBookRevokeDialog.newDialog(
          in_revoke_type, () -> in_books.bookRevoke(in_account, in_book_id));

      final FragmentManager fm = in_activity.getSupportFragmentManager();
      d.show(fm, "revoke-confirm");
    });
  }
}

File: simplified-app-shared/src/main/java/org/nypl/simplified/app/catalog/CatalogBookRevokeButton.kt
package org.nypl.simplified.app.catalog

import android.text.TextUtils
import android.util.TypedValue
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatButton
import org.nypl.simplified.accounts.database.api.AccountType
import org.nypl.simplified.app.R
import org.nypl.simplified.books.api.BookID
import org.nypl.simplified.books.controller.api.BooksControllerType
import java.util.Objects

/**
 * A button for revoking loans or holds.
 */

class CatalogBookRevokeButton(
    activity: AppCompatActivity,
    booksController: BooksControllerType,
    account: AccountType,
    bookID: BookID,
    revokeType: CatalogBookRevokeType)
  : AppCompatButton(activity), CatalogBookButtonType {

  init {
    val resources =
      Objects.requireNonNull(activity.resources)

    when (revokeType) {
      CatalogBookRevokeType.REVOKE_LOAN -> {
        this.text =
          Objects.requireNonNull(resources.getString(R.string.catalog_book_revoke_loan))
        this.contentDescription =
          Objects.requireNonNull(resources.getString(R.string.catalog_accessibility_book_revoke_loan))
      }
      CatalogBookRevokeType.REVOKE_HOLD -> {
        this.text =
          Objects.requireNonNull(resources.getString(R.string.catalog_book_revoke_hold))
        this.contentDescription =
          Objects.requireNonNull(resources.getString(R.string.catalog_accessibility_book_revoke_hold))
      }
    }

    this.ellipsize = TextUtils.TruncateAt.END
    this.maxLines = 1
    this.setTextSize(TypedValue.COMPLEX_UNIT_SP, 12.0f)

    this.setOnClickListener { view ->
      val builder = AlertDialog.Builder(activity)

      when (revokeType) {
        CatalogBookRevokeType.REVOKE_LOAN -> {
          builder.setMessage(R.string.catalog_book_revoke_loan_confirm)
          builder.setPositiveButton(R.string.catalog_book_revoke_loan) { _, _ ->
            booksController.bookRevoke(account, bookID)
          }
        }
        CatalogBookRevokeType.REVOKE_HOLD -> {
          builder.setMessage(R.string.catalog_book_revoke_hold_confirm)
          builder.setPositiveButton(R.string.catalog_book_revoke_hold) { _, _ ->
            booksController.bookRevoke(account, bookID)
          }
        }
      }

      builder.create().show()
    }
  }
}
